File: common/src/main/java/io/xol/chunkstories/net/PacketOutgoingBuffered.java
package io.xol.chunkstories.net;

import java.io.DataOutputStream;
import java.io.IOException;

//(c) 2015-2017 XolioWare Interactive
//http://chunkstories.xyz
//http://xol.io

public class PacketOutgoingBuffered implements PacketOutgoing {
	final PacketsContextCommon context;
	final short id;
	final int payload_size;
	final byte[] payload;

	public PacketOutgoingBuffered(PacketsContextCommon context, short id, int payload_size, byte[] payload) {
		this.context = context;

		this.id = id;
		this.payload_size = payload_size;
		this.payload = payload;
	}

	private void writePacketIdHeader(DataOutputStream out, short id) throws IOException {
		if (id < 127)
			out.writeByte((byte) id);
		else {
			out.writeByte((byte) (0x80 | id >> 8));
			out.writeByte((byte) (id % 256));
		}
	}

	public void write(DataOutputStream out) throws IOException {
		writePacketIdHeader(out, id);
		out.writeInt(payload_size);
		out.write(payload);
	}
}

File: common/src/main/java/xyz/chunkstories/net/PacketOutgoingBuffered.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.net

import xyz.chunkstories.api.net.PacketId
import java.io.DataOutputStream

class PacketOutgoingBuffered(val context: PacketsEncoderDecoder, val id: PacketId, val size: Int, val payload: ByteArray) : PacketOutgoing {

    private fun writePacketIdHeader(out: DataOutputStream) {
        if (id < 127) out.writeByte(id) else {
            out.writeByte(0x80 or (id shr 8))
            out.writeByte(id and 0xFF)
        }
    }

    override fun write(out: DataOutputStream) {
        writePacketIdHeader(out)
        out.writeInt(size)
        out.write(payload)
    }
}